pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -144,23 +147,70 @@ type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
 // the output of the stringSliceFunc to be sorted, as results of LabelValues
 // are not sorted.
 //
-// TODO: Consider running stringSliceFunc calls concurrently
+
+type stringSliceFuncResult struct {
+	Result   []string
+	Warnings storage.Warnings
+}
+
 func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+
+	g, _ := errgroup.WithContext(m.ctx)
+
 	var warnings storage.Warnings
 	resultMap := make(map[string]struct{})
-	for pos, tenantID := range m.tenantIDs {
-		result, resultWarnings, err := f(m.queriers[pos])
-		if err != nil {
-			return nil, nil, err
-		}
-		for _, e := range result {
-			resultMap[e] = struct{}{}
-		}
-		for _, w := range resultWarnings {
-			warnings = append(warnings, fmt.Errorf("error querying tenant id %s: %w", tenantID, w))
+	resultCh := make(chan *stringSliceFuncResult)
```

Comment:
If we make this channel with big enough buffer (`len(m.tenantsID)`), we don't need background goroutine to read from channel. We can then simply read results after `g.Wait()` returns (with no error). It would simplify the method a bit (eg. no need to close `resultCh`, no need for `resultDone` at all).

Reply:
Good point. I have implemented this in a5b94abc6e002f6038d755fc1e9007316b3c3712